### PR TITLE
test_install_asyncio_reactor: restore the original reactor when done

### PR DESCRIPTION
**Pull request by Gallaecio**
Fixes https://github.com/scrapy/scrapy/issues/5765

### CODE REVIEW

**Comment by wRAR** [comment]
I think we shouldn't modify the reactor in the test process, either spawning subprocesses instead (like we already do with multiple CrawlerProcess tests) or doing read-only checks. There were some tests that installed a reactor and so affected later tets, I think they were removed since then. The test that is faling here was, as I mentioned in the issue, initially documented as doing nothing and basically expecting the install function to not raise an exception so we should find why is this no longer what it does and in the worst case remove/rewrite it.

File: scrapy/utils/reactor.py [hunk 1]
```diff
@@ -90,11 +90,16 @@ def set_asyncio_event_loop(event_loop_path):
...
         except RuntimeError:
-            # `get_event_loop` is expected to fail when called from a new thread
+            # Raised by policy.get_running_loop() when called from a new thread
             # with no asyncio event loop yet installed. Such is the case when
-            # called from `scrapy shell`
+            # called from `scrapy shell`.
+            # In Python 3.10, but not in 3.11, policy.get_event_loop() seems to
+            # log a warning when called from the main thread, so we use
+            # asyncio.get_running_loop() to get an earlier RuntimeError as a
+            # workaround.
```
Comment (Gallaecio) [inline scrapy/utils/reactor.py:102]: I think this is happening, it feels like a Python 3.10 bug, but I have found no information about it.

If you comment out `asyncio.get_running_loop()` above, `test_install_asyncio_reactor` fails in Python 3.10 but _not_ in Python 3.11. You can try for yourself with:

```
tox -e py311,py310 -- tests/test_utils_asyncio.py::AsyncioTest tests/test_command_shell.py::ShellTest::test_shell_fetch_async
```

**Comment by Gallaecio** [comment]
So, the current proposal seems to work, but… I don’t understand why. There is a scenario where, in Python 3.10, `asyncio.get_running_loop()` raises `RuntimeError` but `policy.get_event_loop()` does not, yet `policy.get_event_loop()` seems to be the right code to use, as `asyncio.get_running_loop()` does not seem to raise `RuntimeError` for the separate-thread scenario.

**Comment by Gallaecio** [comment]
Also, it [seems](https://github.com/python/cpython/issues/100160#issuecomment-1345525581) they are reverting the warning soon, and re-introducing it later in Python 3.12.

File: scrapy/utils/reactor.py (hunk 1 above)
Comment (Gallaecio) [reply scrapy/utils/reactor.py:102]: My Python 3.11 is 3.11.0, that explains the 3.10-3.11 difference I get locally.

**Comment by Gallaecio** [comment]
Every attempt I was doing to fix the new errors was causing the scrapy shell test to fail until I used both get_running_loop and policy.get_event_loop together. I thought each was raising RuntimeError or not in different scenarios.

However, it turns out the problem was that in some scenarios a different part of the code, outside of this function, was causing the scrapy shell test to fail. Once I realized that, the path to a clean fix became more obvious.

**Comment by Gallaecio** [comment]
Damn, get_running_loop “can only be called from a coroutine or a callback.”

**Review by wRAR — APPROVED** [review]

**Comment by wRAR** [comment]
Great!

**Pull request merged**